src/libsyntax/parse/lexer/mod.rs

```diff
@@ -1445,7 +1445,7 @@ impl<'a> StringReader<'a> {
                                     format!("\"{}\"", &self.src[start..end]),
                                     Applicability::MachineApplicable
                                 ).emit();
-                            return Ok(token::Literal(token::Str_(Symbol::intern("??")), None))
+                            return Ok(token::Literal(token::Char(Symbol::intern("??")), None))
```

Comment:
Can we change `token::Literal(token::Char(_))` to `token::Literal(token::Err(_))` here?

I what I would want to ideally have is the following behavior:

 - When encountering an invalid char literal, we suggest using a string, but we stop type checking anything from that literal.
 - The parser accepts `let x: &str = 'ab';` and `let x: char = 'ab';` _without_ emitting errors (because they are both roughly equally likely to be correct).
 - We continue parsing, parse errors afterwards are emitted. This way if you make this mistake at the beginning of the file you can still address more important issues that come later.
 - We continue type checking if other unrecoverable parse errors don't happen. This way we avoid the "I fixed the 1 parse error and now I have 300 type errors" problem.

I believe all of the above points can be accomplished with the proposed change.

To assure that that is happening I would also want to expand a test roughly as follows:

```
fn main() {
    let x: &str = 'ab';  // only invalid char error
    let y: char = 'cd';  // only invalid char error
    let z = 'ef';  // only invalid char error

    if x == y {}  // no error here
    if y == z {}  // no error here
    if x == z {}  // no error here

    let a: usize = "";  // type error here to confirm we got past the parser
}
```

Does this sound reasonable?

Reply:
This behavior is great, we can see more information.
I guess `fatal_span_verbose` aborts other errors, when I added the test you want, there was one error about `let x: &str = 'ab';` only.
https://github.com/rust-lang/rust/blob/daa53a52a2667533d5fe59bfcc5b8614b79c3d31/src/libsyntax/parse/lexer/mod.rs#L1458-L1460

I think I should use the func except for `fatal_span_verbose` or add `error_span_verbose`.

Reply:
I believe you can replace that with `err_span_` (which uses `struct_span_err` internally), which I think will allow parsing to continue.

Reply:
Hmm, I encounter this error. I want to skip 4th error, any solution?

```
error: character literal may only contain one codepoint
  --> $DIR/lex-bad-char-literals-6.rs:2:19
   |
LL |     let x: &str = 'ab';  //~ ERROR: character literal may only contain one codepoint
   |                   ^^^^

error: character literal may only contain one codepoint
  --> $DIR/lex-bad-char-literals-6.rs:4:19
   |
LL |     let y: char = 'cd';  //~ ERROR: character literal may only contain one codepoint
   |                   ^^^^

error: character literal may only contain one codepoint
  --> $DIR/lex-bad-char-literals-6.rs:6:13
   |
LL |     let z = 'ef';  //~ ERROR: character literal may only contain one codepoint
   |             ^^^^

error: expected `:`, found `;`
  --> $DIR/lex-bad-char-literals-6.rs:2:23
   |
LL |     let x: &str = 'ab';  //~ ERROR: character literal may only contain one codepoint
   |                       ^ expected `:`

error: aborting due to 4 previous errors
```

Reply:
I might been solved this, but new error appeared.
This behavior is bad. rustc parses comment, because `character literal may ...` error doesn't abort the process, I think. So we should add something to judge if there are any comments.
```
error: character literal may only contain one codepoint
  --> $DIR/lex-bad-char-literals-4.rs:4:5
   |
LL |     '●  //~ ERROR: character literal may only contain one codepoint: '●
   |     ^^

error: character literal may only contain one codepoint
  --> $DIR/lex-bad-char-literals-4.rs:4:70
   |
LL |     '●  //~ ERROR: character literal may only contain one codepoint: '●
   |                                                                      ^^

error: expected one of `.`, `;`, `?`, or an operator, found `~`
  --> $DIR/lex-bad-char-literals-4.rs:4:11
   |
LL |     '●  //~ ERROR: character literal may only contain one codepoint: '●
   |           ^ expected one of `.`, `;`, `?`, or an operator here

error: aborting due to 3 previous errors
```

Reply:
I see how that error could arise. I feel we'll have to hack something up here :-|

I think we should bail if you encounter `//` or `\n` and revert to the current behavior in those cases. If you post a WIP commit I can try and point you to the right lines to change.

What does the output of other tests look like?